sinks/attribution/csv.go

```diff
@@ -0,0 +1,67 @@
+package attribution
+
+import (
+	"encoding/csv"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+type tsvField int
+
+const (
+	// the order in which these appear determines the
+	// order of the fields in the resultant TSV
+	TsvName tsvField = iota
+	TsvCardinality
+	TsvDigests
+	TsvOwner
+)
+
+var tsvSchema = [...]string{
+	TsvName:        "Name",
+	TsvCardinality: "Value",
+	TsvDigests:     "Digests",
+	TsvOwner:       "Owner",
+}
+
+// EncodeInterMetricCSV generates a newline-terminated CSV row that describes
+// the data represented by the InterMetric.
+// The caller is responsible for setting w.Comma as the appropriate delimiter.
+// For performance, encodeCSV does not flush after every call; the caller is
+// expected to flush at the end of the operation cycle
+func encodeInterMetricCSV(g *TimeseriesGroup, w *csv.Writer, includeDigests bool) error {
+	digests := ""
+	if includeDigests {
+		var strDigests []string
+		for _, digest := range g.Digests {
+			strDigests = append(strDigests, strconv.FormatUint(uint64(digest), 10))
+		}
+		digests = strings.Join(strDigests, ",")
+	}
+
+	fields := [...]string{
+		TsvName:        g.Name,
+		TsvCardinality: fmt.Sprintf("%d", g.Sketch.Estimate()),
+		TsvDigests:     digests,
```

Comment:
I'm looking for critique here.

The header of the CSV has two self-explanatory fields:
- There is a metric name, as we want to intentionally group timeseries of the same name together -- metrics of distinct timeseries, but the same name, are often emitted from the same line of code, but with a different tags hash.
- There is a metric owner, which at Stripe we want to tie back to an engineering team, but is left intentionally vague in the Veneur source code as this could also be a service, a k8s namespace, etc.

I feel reasonably confident that these two are useful as-is. The other two I am less confident about.

`TsvCardinality` is a count of the unique timeseries associated with a single (metric name, metric owner) in a flush interval. This is useful because we can take CSV reported downstream, and so long as every timeseries emitted from a Veneur sidecar is unique globally, we can sum together the cardinalities per (metric name, metric owner) to get a global view of cardinality per (metric name, metric owner).

However, this assumption does not hold globally. If we can't assume that all metrics are globally unique (e.g. at Stripe, we have a fair amount of metrics that omit the `host` tag altogether, meaning the same timeseries can be flushed to a TSDB such as SignalFx from more than one instance of Veneur), then we can no longer take all the CSVs and sum together by (metric name, metric owner) to get a global view of cardinality.

To support this case, we introduce a field `TsvDigests`, which computes a non-cryptographic hash of every timeseries corresponding to a (metric name, metric owner), then adds it to the corresponding CSV row. This way, whatever downstream service aggregates CSVs together to come up with a global view of cardinality can use a HashSet to perform cardinality computation.

However, there are a few problems with this approach. These are the problems, and what I think we should do about them, and I'm looking for critique from the reviewer(s):

1. In the event that Veneur is configured to produce the `TsvDigests` field, `TsvCardinality` becomes useless because `TsvCardinality` can no longer be trivially summed. To solve this either produce one field or the other, and use the S3 path to denote which version we're using.
2. `TsvDigests` can potentially become long. Every digest is around ~10B long. In practice, we rarely have services at Stripe emit more than >100k MTS per instance - in this case, we'd be adding approx. 10MB (before gzip) per CSV. For 1M MTS (1 order of magnitude higher than we've seen in production), 100MB (before gzip). I think this _feels_ bad (inelegant), but I don't think there's a way we can fix it unless we start introducing additional services (e.g. a centralized service that tallies digests using HLL, before flushing to S3).
3. `TsvDigests` makes CSV rows potentially very long. Software such as Spark may prefer to read something like:

```
metric.a.b storage hash1
metric.a.b storage hash2
metric.a.b storage hash3
```

Rather than:

```
metric.a.b storage hash1,hash2,hash3
```

We could make this change, if we decide to go with a different format like in (1).

Reply:
@imran-stripe and I chatted, and while we agreed that including digests as part of the resultant CSV was inelegant, we were unable to figure out a better solution that did not require adding additional services into the mix.